File: tia-client/src/main/java/com/teamscale/tia.client/JUnit5TestWiseCoverageExecutionListener.java
package com.teamscale.tia.client;

import com.teamscale.report.testwise.model.ETestExecutionResult;
import org.junit.platform.engine.TestExecutionResult;
import org.junit.platform.engine.TestSource;
import org.junit.platform.engine.support.descriptor.ClassSource;
import org.junit.platform.engine.support.descriptor.MethodSource;
import org.junit.platform.launcher.TestExecutionListener;
import org.junit.platform.launcher.TestIdentifier;
import org.junit.platform.launcher.TestPlan;

import java.util.Optional;

/**
 * {@link TestExecutionListener} that uses the {@link TiaAgent} to record test-wise coverage.
 */
@SuppressWarnings("unused")
public class JUnit5TestWiseCoverageExecutionListener implements TestExecutionListener {

	private final RunListenerAgentBridge bridge = new RunListenerAgentBridge(
			JUnit5TestWiseCoverageExecutionListener.class.getName());

	@Override
	public void executionStarted(TestIdentifier testIdentifier) {
		if (!testIdentifier.isTest()) {
			return;
		}
		String uniformPath = getUniformPath(testIdentifier);
		bridge.testStarted(uniformPath);
	}

	private String getUniformPath(TestIdentifier testIdentifier) {
		return testIdentifier.getSource().flatMap(this::parseTestSource).orElse(testIdentifier.getDisplayName());
	}

	private Optional<String> parseTestSource(TestSource source) {
		if (source instanceof ClassSource) {
			ClassSource classSource = (ClassSource) source;
			return Optional.of(classSource.getClassName().replace('.', '/'));
		} else if (source instanceof MethodSource) {
			MethodSource methodSource = (MethodSource) source;
			return Optional.of(methodSource.getClassName().replace('.', '/') + "/" +
					methodSource.getMethodName() + "(" + methodSource.getMethodParameterTypes() + ")");
		}
		return Optional.empty();
	}

	@Override
	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
		if (!testIdentifier.isTest()) {
			return;
		}
		String uniformPath = getUniformPath(testIdentifier);
		ETestExecutionResult result;
		switch (testExecutionResult.getStatus()) {
			case SUCCESSFUL:
				result = ETestExecutionResult.PASSED;
				break;
			case ABORTED:
				result = ETestExecutionResult.ERROR;
				break;
			case FAILED:
			default:
				result = ETestExecutionResult.FAILURE;
				break;
		}

		bridge.testFinished(uniformPath, result);
	}

	@Override
	public void executionSkipped(TestIdentifier testIdentifier, String reason) {
		if (testIdentifier.isContainer()) {
			return;
		}
		String uniformPath = getUniformPath(testIdentifier);
		bridge.testSkipped(uniformPath, reason);
	}

	@Override
	public void testPlanExecutionFinished(TestPlan testPlan) {
		bridge.testRunFinished();
	}
}

File: tia-runlisteners/src/main/kotlin/com/teamscale/tia/runlistener/JUnit5TestwiseCoverageExecutionListener.kt
package com.teamscale.tia.runlistener

import com.teamscale.report.testwise.model.ETestExecutionResult
import com.teamscale.tia.client.TiaAgent
import org.junit.platform.engine.TestExecutionResult
import org.junit.platform.engine.TestSource
import org.junit.platform.engine.support.descriptor.ClassSource
import org.junit.platform.engine.support.descriptor.MethodSource
import org.junit.platform.launcher.TestExecutionListener
import org.junit.platform.launcher.TestIdentifier
import org.junit.platform.launcher.TestPlan
import java.util.*

/**
 * [TestExecutionListener] that uses the [TiaAgent] to record test-wise coverage.
 */
class JUnit5TestwiseCoverageExecutionListener : TestExecutionListener {
	private val bridge = RunListenerAgentBridge.create<JUnit5TestwiseCoverageExecutionListener>()

	override fun executionStarted(testIdentifier: TestIdentifier) {
		if (!testIdentifier.isTest) return
		val uniformPath = testIdentifier.getUniformPath()
		bridge.testStarted(uniformPath)
	}

	private fun TestIdentifier.getUniformPath() =
		source.flatMap { source ->
			source.parse()
		}.orElse(displayName)

	private fun TestSource.parse() =
		when (this) {
			is ClassSource -> Optional.of(className.replace('.', '/'))
			is MethodSource -> Optional.of(
				"${className.replace('.', '/')}/$methodName($methodParameterTypes)"
			)
			else -> Optional.empty()
		}

	override fun executionFinished(testIdentifier: TestIdentifier, testExecutionResult: TestExecutionResult) {
		if (!testIdentifier.isTest) {
			return
		}
		val uniformPath = testIdentifier.getUniformPath()
		val result = when (testExecutionResult.status) {
			TestExecutionResult.Status.SUCCESSFUL -> ETestExecutionResult.PASSED
			TestExecutionResult.Status.ABORTED -> ETestExecutionResult.ERROR
			TestExecutionResult.Status.FAILED -> ETestExecutionResult.FAILURE
			else -> ETestExecutionResult.FAILURE
		}

		bridge.testFinished(uniformPath, result)
	}

	override fun executionSkipped(testIdentifier: TestIdentifier, reason: String) {
		if (testIdentifier.isContainer) return
		val uniformPath = testIdentifier.getUniformPath()
		bridge.testSkipped(uniformPath, reason)
	}

	override fun testPlanExecutionFinished(testPlan: TestPlan) {
		bridge.testRunFinished()
	}
}
